### PR TITLE
Add validateValueAtPath before setting properties

### LINKED ISSUE #114: Instance property must be an Array, not an Object
Elements that are defined as 0..* in resources must be JSON arrays in the instance, even if the cardinality has been constrained, or the array contains just one value. It appears array elements are being rendered as Objects rather than Arrays.

Example -- on master branch, at commit `https://github.com/standardhealth/fsh-mcode/commit/4948b6ebe0018ddf2f9d8d8c53a469de972f059d`

Example: `build\input\resources\Condition-mCODEComorbidConditionExample01.json`

IG Publisher reports:  `Condition.category | error | This property must be an Array, not an Object`

### PR DESCRIPTION
**Pull request by jafeltra**
Fixes #114 

The functionality for putting properties into arrays even if they are constrained is in the `validateValueAtPath` function. We call that before setting properties on instances that are defined by rules. However, we also set properties on instances based on the structure definition. I updated the function `setFixedValuesForDirectChildren` to also call `validateValueAtPath`. This will check the base cardinality and should fix the bug. I added two tests to check the two places the `setFixedValuesForDirectChildren` is called and check that the arrays are properly set.

### CODE REVIEW

File: test/export/InstanceExporter.test.ts [hunk 1]
```diff
@@ -301,6 +325,26 @@ describe('InstanceExporter', () => {
       expect(exported.communication).toBeUndefined();
     });
 
+    it('should fix a nested element that has parents defined in the instance and fixed on the SD to an array even if constrained', () => {
+      const cardRule = new CardRule('communication');
```
Comment (mint-thompson) [inline test/export/InstanceExporter.test.ts]: Should this `CardRule` apply to `contact` here? The test still passes when this change is made.

File: test/export/InstanceExporter.test.ts [hunk 2]
```diff
@@ -185,6 +185,30 @@ describe('InstanceExporter', () => {
...
+      const condition = new Profile('TestCondition');
+      condition.parent = 'Condition';
+      doc.profiles.set(condition.name, condition);
+      const conditionInstance = new Instance('Bar');
+      conditionInstance.instanceOf = 'TestCondition';
+      doc.instances.set(conditionInstance.name, conditionInstance);
+      const fixedValRule = new FixedValueRule('category');
+      const fixedFshCode = new FshCode('foo', 'http://foo.com');
+      fixedValRule.fixedValue = fixedFshCode;
+      condition.rules.push(fixedValRule);
```
Comment (mint-thompson) [inline test/export/InstanceExporter.test.ts]: Should `condition` have a `CardRule` added to constrain `category` to `1..1` here?

File: src/export/InstanceExporter.ts [hunk 3]
```diff
@@ -76,8 +89,9 @@ export class InstanceExporter {
    */
   private setFixedValuesForDirectChildren(
     element: ElementDefinition,
-    existingPath: PathPart[],
-    instanceDef: InstanceDefinition
+    existingPath: string,
+    instanceDef: InstanceDefinition,
+    instanceOfStructureDefinition: StructureDefinition
```
Comment (mint-thompson) [inline src/export/InstanceExporter.ts]: Since the function arguments changed, could you also updated the `@param` listing above the function?

File: test/export/InstanceExporter.test.ts (hunk 1 above)
Comment (jafeltra) [reply test/export/InstanceExporter.test.ts]: Oops yes! Good catch!

File: test/export/InstanceExporter.test.ts (hunk 2 above)
Comment (jafeltra) [reply test/export/InstanceExporter.test.ts]: For this one, I was trying to mirror the example that Mark mentioned in the original issue. It didn't look like that example was constraining the cardinality. It only has the rule `* category = SCT#398192003`, which I think we interpret as setting the value to the code (an object). When I undid my changes to the exporter, this test failed, so it seemed valid to me, but I'd be open to adding the `CardRule`. I was just trying to keep it as similar to the example as possible.

Do you have thoughts on which way to do this? I'm open to either.

**Comment by jafeltra** [comment]
As for your top level comment about the string manipulation, if you have ideas about how to make it more clear, I'm happy to change it. From what I could see, I had to rebuild the string in that place because parsing the `rule.path` into the `PathPart[]` adds in extra indices to the path that weren't originally there, and I needed a string representation for the `validateValueAtPath` function to work properly since it outputs the path string in the error. If you think I should change/improve the manipulation, let me know!

File: test/export/InstanceExporter.test.ts (hunk 2 above)
Comment (cmoesel) [reply test/export/InstanceExporter.test.ts]: I doubt it makes a difference, but Mark's example is for a profile whose parent is USCoreCondition.  USCoreCondition constrains `category` to `1..*`.  So you could add that constraint if you want to be as close as possible -- although I doubt it will make a difference.

**Comment by cmoesel** [comment]
I tried to think of a better way than the string rebuilding too, but couldn't come up with anything.  I think you've found the best way.  If there's something better, it probably involves a lot of refactoring.

So... if both of you are good with this, then I'm good with it!

File: test/export/InstanceExporter.test.ts (hunk 2 above)
Comment (jafeltra) [reply test/export/InstanceExporter.test.ts]: Ah good point! I just added it (locally) and it still passes and fails correctly, so I think it makes sense to add it. I can push that!

**Comment by jafeltra** [comment]
I wasn't able to think of anything better than the string manipulation, so I'm good with that part if Mint is!

The new errors you are getting are because we are now validating the paths before setting the properties, so the errors from issue #121 are showing up now. I wrapped the calls for validating and setting in a try/catch so that the instance export will continue despite the error. You should now get the 22 instances, but they'll be missing the invalid slice. The last two commits add that try/catch and the `CardRule` to the test.

However, Chris also noticed that there is a value missing in the `mCODECancerRelatedRadiationProcedureExample01`, so I'm also going to look into why that's happening.

File: src/export/InstanceExporter.ts (hunk 3 above)
Comment (mint-thompson) [reply src/export/InstanceExporter.ts]: Ah, I should have been more specific: there's also the new `instanceOfStructureDefinition` that should be added to the `@param` listing.

**Comment by jafeltra** [comment]
After an eventful afternoon, @cmoesel  and @kjmahalingam helped me figure out why the fixed value was missing from `mCODECancerRelatedRadiationProcedureExample01`. The root of the issue was the order in which we applied values to instances, so I swapped around the order we do this in. We now first set values from the Structure Definition the instance instantiates as necessary and then set values that are fixed directly on the instance. I also added the missing param that @mint-thompson pointed out.

I've compared the diffs between exporting `fsh-mcode` on this branch and on `master` and the only changes are properties in different orders (given the switch in how we export now) and the missing invalid slices (because we now skip over them instead of adding invalid slices, and we will fix the slices in issue #121).

In conclusion, I _think_ this branch should be set now.
